File: swift/bob/Sources/Bob/Bob.swift
import Foundation

extension String {   
    func trim() -> String {
        return self.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

func hey(_ input: String) -> String {
    let trimmed = input.trim()

    // silence
    if trimmed.isEmpty { return "Fine. Be that way!" }

    // shouting: input contains letters but no lowercase letters
    let shouting: Bool = 
        // trimmed.rangeOfCharacter(from: .letters) != nil &&
        // trimmed.rangeOfCharacter(from: .lowercaseLetters) == nil
        trimmed == trimmed.uppercased() &&
        trimmed != trimmed.lowercased()
    if shouting { return "Whoa, chill out!" }

    // ask a question
    if trimmed.hasSuffix("?") { return "Sure." } 

    // anything else
    return "Whatever."
}

File: euphoria/bob/bob.ex
include std/regex.e as re
include std/text.e

public function hey(sequence input)
  sequence trimmed = trim(input)

  atom isSilent = length(trimmed) = 0
  atom isAsking = re:has_match(re:new("[?]$"), trimmed)
  atom isYelling = re:has_match(re:new("[[:upper:]]"), trimmed)
                   and not re:has_match(re:new("[[:lower:]]"), trimmed)

  if isSilent then
    return "Fine. Be that way!"
  elsif isAsking and isYelling then
    return "Calm down, I know what I'm doing!"
  elsif isAsking then
    return "Sure."
  elsif isYelling then
    return "Whoa, chill out!"
  else
    return "Whatever."
  end if
end function
